File: rental/src/main/java/io/axoniq/demo/bikerental/rental/command/Bike.java
package io.axoniq.demo.bikerental.rental.command;

import io.axoniq.demo.bikerental.coreapi.rental.ApproveRequestCommand;
import io.axoniq.demo.bikerental.coreapi.rental.BikeInUseEvent;
import io.axoniq.demo.bikerental.coreapi.rental.BikeRegisteredEvent;
import io.axoniq.demo.bikerental.coreapi.rental.BikeRequestedEvent;
import io.axoniq.demo.bikerental.coreapi.rental.BikeReturnedEvent;
import io.axoniq.demo.bikerental.coreapi.rental.RegisterBikeCommand;
import io.axoniq.demo.bikerental.coreapi.rental.RejectRequestCommand;
import io.axoniq.demo.bikerental.coreapi.rental.RequestBikeCommand;
import io.axoniq.demo.bikerental.coreapi.rental.RequestRejectedEvent;
import io.axoniq.demo.bikerental.coreapi.rental.ReturnBikeCommand;
import org.axonframework.commandhandling.CommandHandler;
import org.axonframework.eventsourcing.EventSourcingHandler;
import org.axonframework.modelling.command.AggregateIdentifier;
import org.axonframework.spring.stereotype.Aggregate;

import java.util.Objects;
import java.util.UUID;

import static org.axonframework.modelling.command.AggregateLifecycle.apply;

@Aggregate
public class Bike {

    @AggregateIdentifier
    private String bikeId;

    private boolean isAvailable;
    private String reservedBy;
    private boolean reservationConfirmed;

    public Bike() {
    }

    @CommandHandler
    public Bike(RegisterBikeCommand command) {
        apply(new BikeRegisteredEvent(command.bikeId(), command.bikeType(), command.location()));
    }

    @CommandHandler
    public String handle(RequestBikeCommand command) {
        if (!this.isAvailable) {
            throw new IllegalStateException("Bike is already rented");
        }
        String rentalReference = UUID.randomUUID().toString();
        apply(new BikeRequestedEvent(command.bikeId(), command.renter(), rentalReference));

        return rentalReference;
    }

    @CommandHandler
    public void handle(ApproveRequestCommand command) {
        if (!Objects.equals(reservedBy, command.renter())
                || reservationConfirmed) {
            return;
        }
        apply(new BikeInUseEvent(command.bikeId(), command.renter()));
    }

    @CommandHandler
    public void handle(RejectRequestCommand command) {
        if (!Objects.equals(reservedBy, command.renter())
                || reservationConfirmed) {
            return;
        }
        apply(new RequestRejectedEvent(command.bikeId()));
    }

    @CommandHandler
    public void handle(ReturnBikeCommand command) {
        if (this.isAvailable) {
            throw new IllegalStateException("Bike was already returned");
        }
        apply(new BikeReturnedEvent(command.bikeId(), command.location()));
    }

    @EventSourcingHandler
    protected void handle(BikeRegisteredEvent event) {
        this.bikeId = event.bikeId();
        this.isAvailable = true;
    }

    @EventSourcingHandler
    protected void handle(BikeReturnedEvent event) {
        this.isAvailable = true;
        this.reservationConfirmed = false;
        this.reservedBy = null;
    }

    @EventSourcingHandler
    protected void handle(BikeRequestedEvent event) {
        this.reservedBy = event.renter();
        this.reservationConfirmed = false;
        this.isAvailable = false;
    }

    @EventSourcingHandler
    protected void handle(RequestRejectedEvent event) {
        this.reservedBy = null;
        this.reservationConfirmed = false;
        this.isAvailable = true;
    }

    @EventSourcingHandler
    protected void on(BikeInUseEvent event) {
        this.isAvailable = false;
        this.reservationConfirmed = true;
    }
}

File: rental/src/main/kotlin/io/axoniq/demo/bikerental/rental/command/Bike.kt
package io.axoniq.demo.bikerental.rental.command

import io.axoniq.demo.bikerental.coreapi.rental.*
import org.axonframework.commandhandling.CommandHandler
import org.axonframework.eventsourcing.EventSourcingHandler
import org.axonframework.modelling.command.AggregateIdentifier
import org.axonframework.modelling.command.AggregateLifecycle
import org.axonframework.spring.stereotype.Aggregate
import java.util.*

@Aggregate
class Bike {

    @AggregateIdentifier
    private var bikeId: String? = null

    private var isAvailable = false
    private var reservedBy: String? = null

    constructor()

    @CommandHandler
    constructor(command: RegisterBikeCommand) {
        AggregateLifecycle.apply(BikeRegisteredEvent(command.bikeId, command.bikeType, command.location))
    }

    @CommandHandler
    fun handle(command: RequestBikeCommand): String {
        check(this.isAvailable) { "Is already requested!" }

        val event = BikeRequestedEvent(command.bikeId, command.renter, rentalReference = UUID.randomUUID().toString())
        AggregateLifecycle.apply(event)

        return event.rentalReference
    }

    @CommandHandler
    fun handle(command: ApproveRequestCommand) {
        if (command.renter != reservedBy) {
            return
        }

        AggregateLifecycle.apply(BikeInUseEvent(command.bikeId, command.renter))
    }

    @CommandHandler
    fun handle(command: RejectRequestCommand) {
        if (command.renter != reservedBy) {
            return
        }

        AggregateLifecycle.apply(RequestRejectedEvent(command.bikeId))
    }

    @CommandHandler
    fun handle(command: ReturnBikeCommand) {
        check(!this.isAvailable) { "Bike was already returned" }
        AggregateLifecycle.apply(BikeReturnedEvent(command.bikeId, command.location))
    }

    @EventSourcingHandler
    private fun handle(event: BikeRegisteredEvent) {
        this.bikeId = event.bikeId
        this.isAvailable = true
    }

    @EventSourcingHandler
    private fun handle(event: BikeRequestedEvent) {
        this.reservedBy = event.renter
        this.isAvailable = false
    }

    @EventSourcingHandler
    private fun handle(event: BikeReturnedEvent) {
        this.reservedBy = null
        this.isAvailable = true
    }

    @EventSourcingHandler
    private fun handle(event: RequestRejectedEvent) {
        this.reservedBy = null
        this.isAvailable = true
    }

    @EventSourcingHandler
    private fun handle(event: BikeInUseEvent) {
        this.isAvailable = false
    }
}
